attendispeechservice/src/main/kotlin/nl/attendi/attendispeechservice/audio/AudioRecorderImpl.kt

```diff
@@ -74,40 +88,55 @@ internal object AudioRecorderImpl : AudioRecorder {
 
             try {
                 @SuppressLint("MissingPermission")
-                audioRecord = AudioRecord(
+                val audioRecord = AudioRecord(
                     audioRecordingConfig.audioSource,
                     audioRecordingConfig.sampleRate,
                     audioRecordingConfig.channel,
                     audioRecordingConfig.encoding,
                     bufferSize
-                ).apply {
-                    startRecording()
-                }
+                )
+                audioRecord.startRecording()
             } catch (e: IllegalArgumentException) {
-                throw AudioRecorderException.UnsupportedAudioFormat(e.message ?: "Invalid Audio Format")
+                throw AudioRecorderException.UnsupportedAudioFormat(
+                    e.message ?: "Invalid Audio Format"
+                )
+            } catch (_: IllegalStateException) {
+                throw AudioRecorderException.DeniedRecodingPermission
             }
 
-            isRecordingInternal = true
-        }
+            // Create channel for audio frames.
+            audioFrameChannel = Channel(Channel.UNLIMITED)
+
+            // Start processing coroutine sequentially.
+            audioProcessingJob = CoroutineScope(Dispatchers.Default).launch {
+                if (audioFrameChannel != null) {
+                    for (frame in audioFrameChannel) {
+                        onAudio(frame)
+                    }
+                }
+            }
 
-        try {
-            while (true) {
-                /**
-                 * Ensures that job in the current context is active.
-                 * If the job is no longer active, throws CancellationException and breaks the while loop.
-                 */
-                coroutineContext.ensureActive()
-                val readSize =
-                    audioRecord?.read(temporaryAudioBuffer, 0, temporaryAudioBuffer.size) ?: 0
-                if (readSize <= 0) {
-                    continue
+
+            // Launch reading coroutine (reads from mic, pushes to channel)
+            audioReadingJob = CoroutineScope(Dispatchers.IO).launch {
+                try {
+                    while (isActive) {
+                        val readSize =
+                            audioRecord?.read(temporaryAudioBuffer, 0, temporaryAudioBuffer.size)
+                                ?: 0
+                        if (readSize <= 0) {
+                            continue
+                        }
+                        val samples = temporaryAudioBuffer.toList().take(readSize)
+                        val audioFrame = AudioFrame(samples)
+                        onAudio(audioFrame)
```

Comment:
It seems like you're never pushing to the channel. Shouldn't this be
```suggestion
                audioFrameChannel?.send(AudioFrame(samples))
```
?